crates/spfs-cli/cmd-fuse/src/cmd_fuse.rs

```diff
@@ -268,11 +294,52 @@ impl CmdFuse {
                 _ = terminate.recv() => Err(miette!("Terminate signal received, filesystem shutting down")),
                 _ = interrupt.recv() => Err(miette!("Interrupt signal received, filesystem shutting down")),
                 _ = quit.recv() => Err(miette!("Quit signal received, filesystem shutting down")),
+                _ = heartbeat_recv.recv() => {
+                    heartbeat_failed = true;
+                    Err(miette!("Heartbeat monitor triggered, filesystem shutting down"))
+                }
             };
-            // the filesystem task must be fully terminated in order for the subsequent unmount
+
+            if let Some(handle) = heartbeat_monitor.take() {
+                handle.abort_handle().abort();
+                if let Err(err) = handle.await {
+                    tracing::warn!("Heartbeat monitor failed: {err:?}");
+                }
+            }
+
+            // The filesystem task must be fully terminated in order for the subsequent unmount
             // process to function. Otherwise, the background task will keep this process alive
             // forever.
-            abort_handle.abort();
+            //
+            // Exception: if spfs-monitor died without cleaning up the runtime,
+            // fuser::Session::run will not exit and there's no way to tell it
+            // to break out of its loop. When the fuse filesystem is included
+            // in the lowerdir of an overlayfs, even though `fusermount -u`
+            // appears to succeed, and the mount disappears from /etc/mtab,
+            // spfs-fuse stays alive and still functions when accessed via the
+            // overlayfs. As a last-ditch effort to avoid leaving a spfs-fuse
+            // process running forever, we will return without attempting any
+            // cleanup.
+            if heartbeat_failed {
+                return res;
+            }
+            if !join_handle.is_finished() {
+                // XXX: Calling unmount has no apparent effect!
+                unmount_callable.lock().unwrap().unmount().into_diagnostic().wrap_err("FUSE unmount failed")?;
+                tracing::trace!("Joining FUSE session");
+                // Since the umount above may have no effect, this join uses a
+                // timeout so the process doesn't deadlock. Once this process
+                // exits, the fusermount auto_unmount should kick in.
+                match timeout(Duration::from_secs(5), join_handle).await {
```

Comment:
Is it worth being able to configure this timeout?

Reply:
From my testing no amount of waiting will ever succeed, so making this configurable might just mislead people into thinking it could matter. Except maybe to let it be configured shorter than 5 seconds?

I wanted to leave this code that attempts to clean up gracefully in place although it isn't working properly, and the timeout is a concession to leave the code in place without hanging forever. Hopefully this can be revisited and figured out later.